streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -95,6 +97,7 @@
     private static final String JMX_PREFIX = "kafka.streams";
     public static final int DEFAULT_CLOSE_TIMEOUT = 0;
     private final StreamThread[] threads;
+    private final Map<Long, StreamThread.State> threadState;
```

Comment:
Hmmm... Even if we use `ConcurrentHashMap` I guess this would not be enough. In `StreamStateListener#onChange()` we check the state of each thread to compute the new state for KS -- this most be atomic. Using `ConcurrentHashMap` would not prevent a race condition there.

So maybe `StreamStateListener#onChange()` should have `synchronized` keyword.